### PR TITLE
ceph.spec.in,debian/rules: change aio-max-nr to 1048576

### PR DESCRIPTION
**Pull request by chenliuzhong**
when osd is more than 14 in one host,it report error that aio is not enough.
As the default aio-max-nr is 65536, one OSD needs 4096 aios and other programs may use aios.
This patch change aio-max-nr to 1048576 when install ceph-common rpm package

Signed-off-by: chenliuzhong <liuzhong.chen@easystack.cn>

### CODE REVIEW

**Comment by chenliuzhong** [comment]
@liewegas need a review please.

**Comment by liewegas** [comment]
@b-ranto is this a reasonable approach to adjusting this limit?  If so, we'll want to update the debian packages too.

**Comment by liewegas** [comment]
@chenliuzhong do you mind making a similar change to the debian ceph-osd package too?

**Comment by chenliuzhong** [comment]
@liewegas ok,I will try to make a similar change to the debian package

**Comment by chenliuzhong** [comment]
@liewegas Is there any official docs of  make a  ceph apt package?I want to make apt package to test the change.I try by the mothed online but failed.

**Comment by chenliuzhong** [comment]
@liewegas Is there any official docs of make a ceph apt package?I want to make apt package to test the change.I try by the mothed online but failed.Thank you!

**Comment by tchaikov** [comment]
@chenliuzhong https://github.com/ceph/ceph/blob/master/doc/install/build-ceph.rst#advanced-package-tool-apt

**Comment by chenliuzhong** [comment]
@tchaikov Thank you!

File: ceph.spec.in [hunk 1]
```diff
@@ -884,6 +884,8 @@ mkdir -p %{buildroot}%{_sbindir}
 install -m 0644 -D src/logrotate.conf %{buildroot}%{_sysconfdir}/logrotate.d/ceph
 chmod 0644 %{buildroot}%{_docdir}/ceph/sample.ceph.conf
 install -m 0644 -D COPYING %{buildroot}%{_docdir}/ceph/COPYING
+#modify aio-max-nr =1048576
+install -m 0644 -D src/90-cephsysctl.conf %{buildroot}%{_prefix}/lib/sysctl.d/90-cephsysctl.conf
```
Comment (liewegas) [inline ceph.spec.in]: Can we call this 90-ceph-osd-sysctl.conf, and put it in the ceph-osd package?

Comment (liewegas) [reply ceph.spec.in]: then it's just a matter of adding a line to debian/ceph-osd.install for usr/lib/sysctl.d/90-ceph-osd-sysctl.conf

**Comment by chenliuzhong** [comment]
@liewegas  need review,please.
I have change the modify from ceph-common to ceph-osd and make a similar change to the debian ceph-osd package.

File: ceph.spec.in [hunk 2]
```diff
@@ -884,6 +884,8 @@ mkdir -p %{buildroot}%{_sbindir}
 install -m 0644 -D src/logrotate.conf %{buildroot}%{_sysconfdir}/logrotate.d/ceph
 chmod 0644 %{buildroot}%{_docdir}/ceph/sample.ceph.conf
 install -m 0644 -D COPYING %{buildroot}%{_docdir}/ceph/COPYING
+#modify aio-max-nr =1048576
+install -m 0644 -D src/90-ceph-osd-sysctl.conf %{buildroot}%{_prefix}/lib/sysctl.d/90-ceph-osd-sysctl.conf
```
Comment (tchaikov) [inline ceph.spec.in]: use the macro of `%{_sysctldir}`. see https://fedoraproject.org/wiki/Packaging:Guidelines#binfmt.d.2C_sysctl.d_and_tmpfiles.d

File: ceph.spec.in [hunk 3]
```diff
@@ -1432,6 +1435,8 @@ fi
 if [ $1 -eq 1 ] ; then
 /usr/bin/systemctl start ceph-osd.target >/dev/null 2>&1 || :
 fi
+#make the /usr/lib/sysctl.d/90-ceph-osd-sysctl.conf work
+sysctl -p %{_prefix}/lib/sysctl.d/90-ceph-osd-sysctl.conf
```
Comment (tchaikov) [inline ceph.spec.in]: please and call `%sysctl_apply` and use the `sysctl -p` as a fallback, as this macro is not available in el7.0 but exists in el7.2 and up. also, please use full path of binary, like
```
%if 0%{?sysctl_apply}
    %sysctl_apply 90-ceph-osd-sysctl.conf
%else
    /usr/lib/systemd/systemd-sysctl %{_sysctldir}/90-ceph-osd-sysctl.conf >/dev/null 2>&1 || :
%endif
```

File: src/90-ceph-osd-sysctl.conf [hunk 4]
```diff
@@ -0,0 +1 @@
+fs.aio-max-nr = 1048576
```
Comment (tchaikov) [inline src/90-ceph-osd-sysctl.conf]: could you consider using `90-ceph-osd.conf` as the filename? as sysctl is redundant here.

File: ceph.spec.in [hunk 5]
```diff
@@ -884,6 +884,8 @@ mkdir -p %{buildroot}%{_sbindir}
 install -m 0644 -D src/logrotate.conf %{buildroot}%{_sysconfdir}/logrotate.d/ceph
 chmod 0644 %{buildroot}%{_docdir}/ceph/sample.ceph.conf
 install -m 0644 -D COPYING %{buildroot}%{_docdir}/ceph/COPYING
+#modify aio-max-nr =1048576
```
Comment (tchaikov) [inline ceph.spec.in]: please drop this comment.

**Comment by ddiss** [comment]
Change looks good to me. Any chance we could piggy back on this PR to also add:
`kernel.pid_max = 4194303`

The 32768 default (on SLE, not sure about RHEL, Ubuntu, etc. defaults) is pretty easy to hit during recovery on high-osd-count nodes.

File: ceph.spec.in (hunk 5 above)
Comment (chenliuzhong) [reply ceph.spec.in]: ok,I will drop the comment,thank you!

File: debian/rules [hunk 6]
```diff
@@ -50,6 +50,7 @@ override_dh_auto_install:
 	install -D -m 644 udev/95-ceph-osd.rules $(DESTDIR)/lib/udev/rules.d/95-ceph-osd.rules
 	install -D -m 644 udev/60-ceph-by-parttypeuuid.rules $(DESTDIR)/lib/udev/rules.d/60-ceph-by-parttypeuuid.rules
 	install -D -m 644 src/etc-rbdmap $(DESTDIR)/etc/ceph/rbdmap
+	install -D -m 644 src/90-ceph-osd-sysctl.conf $(DESTDIR)/usr/lib/sysctl.d/90-ceph-osd-sysctl.conf
```
[the remark this answers is not in the document]
Comment (chenliuzhong) [reply debian/rules]: OK,I will modify this,thank you!

File: ceph.spec.in [hunk 7]
```diff
@@ -1432,6 +1435,8 @@ fi
 if [ $1 -eq 1 ] ; then
 /usr/bin/systemctl start ceph-osd.target >/dev/null 2>&1 || :
 fi
+#make the /usr/lib/sysctl.d/90-ceph-osd-sysctl.conf work
```
[the remark this answers is not in the document]
Comment (chenliuzhong) [reply ceph.spec.in]: OK, I will drop the comment,thank you!

File: src/90-ceph-osd-sysctl.conf (hunk 4 above)
Comment (chenliuzhong) [reply src/90-ceph-osd-sysctl.conf]: OK，I will change the name,thank you!

File: ceph.spec.in (hunk 3 above)
Comment (chenliuzhong) [reply ceph.spec.in]: ok,I will change like this

Comment (chenliuzhong) [reply ceph.spec.in]: compare to sysctl -p，sysctl_apply has any advantage？

**Comment by smithfarm** [comment]
@chenliuzhong Please rebase - there is a merge conflict.

**Comment by chenliuzhong** [comment]
@smithfarm  thank you for reminding,I will slove it.

File: debian/rules (hunk 6 above)
[the remark this answers is not in the document]
Comment (chenliuzhong) [reply debian/rules]: As the other install repeat the name just like :
udev/60-ceph-by-parttypeuuid.rules $(DESTDIR)/lib/udev/rules.d/60-ceph-by-parttypeuuid.rules,
I think should keep the same form.

File: ceph.spec.in (hunk 2 above)
Comment (chenliuzhong) [reply ceph.spec.in]: OK,I change it

File: debian/ceph-osd.postinst [hunk 8]
```diff
@@ -14,6 +14,8 @@ set -e
 # 	deconfigured's-postinst abort-deconfigure in-favour <failed-install-package> <version> [<removing conflicting-package> <version>]
 #
 
+#make aio-max-nr =1048576 work
+sysctl -p /usr/lib/sysctl.d/90-ceph-osd-sysctl.conf
```
[the remark this answers is not in the document]
Comment (chenliuzhong) [reply debian/ceph-osd.postinst]: ok,I change it like this,thank you

File: ceph.spec.in [hunk 9]
```diff
@@ -1432,6 +1434,11 @@ fi
 if [ $1 -eq 1 ] ; then
 /usr/bin/systemctl start ceph-osd.target >/dev/null 2>&1 || :
 fi
+%if 0%{?sysctl_apply}
+    %sysctl_apply 90-ceph-osd.conf
+%else
+    /usr/lib/systemd/systemed-sysctl %{_sysctldir}/90-ceph-osd.conf > /dev/null 2>&1 || :
+%endif
```
Comment (smithfarm) [inline ceph.spec.in]: here we have "90-ceph-osd.conf" but elsewhere in the same PR it is "30-ceph-osd.conf" - why the difference?

File: debian/rules [hunk 10]
```diff
@@ -50,6 +50,7 @@ override_dh_auto_install:
 	install -D -m 644 udev/95-ceph-osd.rules $(DESTDIR)/lib/udev/rules.d/95-ceph-osd.rules
 	install -D -m 644 udev/60-ceph-by-parttypeuuid.rules $(DESTDIR)/lib/udev/rules.d/60-ceph-by-parttypeuuid.rules
 	install -D -m 644 src/etc-rbdmap $(DESTDIR)/etc/ceph/rbdmap
+	install -D -m 644 src/90-ceph-osd.conf $(DESTDIR)/etc/sysctl.d/30-ceph-osd.conf
```
Comment (smithfarm) [inline debian/rules]: why the difference "90-ceph-osd.conf" -> "30-ceph-osd.conf" ?

Comment (chenliuzhong) [reply debian/rules]: Here  I follow the /etc/sysctl.d/README  of ubuntu : "In general, files in the 10-*.conf range come from the procps package and serve as system defaults.  Other packages install their files in the
30-*.conf range, to override system defaults".
But in Centos I should use 90-*，so change 90-* to 30-* here.

**Comment by tchaikov** [comment]
thanks @chenliuzhong! needs rebase?

**Comment by chenliuzhong** [comment]
@tchaikov  I think there is no need to rebase,but I merge the patch from master careless when I deal with that conflict.can you tell how to delete the patch?

**Comment by smithfarm** [comment]
@chenliuzhong Something like this (assuming your remote is called "ceph") in the PR branch checkout should do the trick:

```
git fetch ceph
git rebase ceph/master
git push -f
```

**Comment by chenliuzhong** [comment]
@smithfarm  thanks,I will slove this issue

**Comment by chenliuzhong** [comment]
@tchaikov @smithfarm   OK,I have slove that, and now this patch is what I want to pull,Thank you for all!

**Comment by chenliuzhong** [comment]
@ddiss   do you mean it should set the aio-max-nr = 1073741824 like SUSE/DeepSea! I calculate it aio-max-nr = 1073741824 can support 262144 OSD per host,and aio-max-nr = 1048576 can support 256 osd per host,and I think 256 is enough. aio-max-nr = 1048576 I refer to libvirt.

**Comment by tchaikov** [comment]
@chenliuzhong i think what @ddiss suggests is to add `kernel.pid_max = 4194303` in the `.conf` file.

**Comment by chenliuzhong** [comment]
@tchaikov @ddiss  I suggest to open a new pull to add kernel.pid_max = 4194303 and tell the reason to do so.

File: ceph.spec.in [hunk 11]
```diff
@@ -1438,6 +1440,11 @@ fi
 if [ $1 -eq 1 ] ; then
 /usr/bin/systemctl start ceph-osd.target >/dev/null 2>&1 || :
 fi
+%if 0%{?sysctl_apply}
+    %sysctl_apply 90-ceph-osd.conf
+%else
+    /usr/lib/systemd/systemed-sysctl %{_sysctldir}/90-ceph-osd.conf > /dev/null 2>&1 || :
```
Comment (smithfarm) [inline ceph.spec.in]: typo - should by "systemd-sysctl" instead of "systemed-sysctl"

File: debian/rules (hunk 10 above)
Comment (smithfarm) [reply debian/rules]: ACK

File: ceph.spec.in (hunk 11 above)
Comment (chenliuzhong) [reply ceph.spec.in]: @smithfarm  yeah,you are right,I have check the code in the server,I will modify it right.Thank you very much.

**Comment by ddiss** [comment]
@chenliuzhong : okay, I'll open a new kernel.pid_max pull request once your change is merged - thanks

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-10-24_16:44:55-rados-wip-kefu-testing-2017-10-24-2303-distro-basic-mira/

**Comment by liewegas** [comment]
luminous backport https://github.com/ceph/ceph/pull/18540

**Pull request merged**